buttercup.el

```diff
@@ -1106,15 +1107,35 @@ responsibility to ensure ARG is a command."
 (defun buttercup--make-spy (fun)
   "Create a new spy function wrapping FUN and tracking calls to itself."
   (let (this-spy-function)
-    (setq this-spy-function
-          (lambda (&rest args)
-            (let ((return-value (apply fun args)))
+    (setq
+     this-spy-function
+     (lambda (&rest args)
+       (let ((returned nil)
+             (return-value nil))
+         (condition-case err
+             (progn
+               (setq return-value (apply fun args)
+                     returned t)
+               (buttercup--spy-calls-add
+                this-spy-function
+                (make-spy-context :args args
+                                  :return-value return-value
+                                  :thrown-signal nil
+                                  :current-buffer (current-buffer)))
+               return-value)
+           (error
+            ;; If returned is non-nil, then the error we caught
+            ;; didn't come from FUN, so we shouldn't record it.
+            (unless returned
               (buttercup--spy-calls-add
                this-spy-function
                (make-spy-context :args args
-                                 :return-value return-value
-                                 :current-buffer (current-buffer)))
-              return-value)))
+                                 :return-value nil
```

Comment:
`nil` could be a valid return value. Tests could misstake that for valid returns unless the test is careful to check for thrown signals.
I suppose we could use a symbol like `buttercup-noreturn' or something.  It seems unlikely that a function outside buttercup would return that value in normal circumstances.

Reply:
There's always the option to use `gensym` to make an unique symbol that's guaranteed not to be in use anywhere else.

Reply:
That sounds like a better idea.  You should only use the `thrown-signal` slot to determine whether a call has thrown an error.

Reply:
Returning any arbitrary symbol isn't great either. If the caller just checks whether the return value is non-nil, the test will pass. With the current design, the onus is entirely on the caller to check the thrown signal in order to determine whether the function failed.

Would it be reasonable for `spy-context-return-value` to throw an error when called on a spy context for a function call that threw an error? (And is this implementable on top of the `cl-defstruct`?)

Reply:
Do you think it would work to use different structs for calls that return and calls that throw?
```emacs-lisp
(cl-defstruct spy-context args current-buffer)
(cl-defstruct (spy-context-return :include spy-context) return-value)
(cl-defstruct (spy-context-error :include spy-context) thrown-signal)
```
Then it would not be possible to get the `return-value` of a call that threw an error. No idea if that has weird consequences somewhere.

As an aside you might want to add functions `spy-calls-count-returned` and  `spy-calls-count-errors` or similar.

Reply:
I think that is, practically speaking, equivalent to my proposal to have `spy-context-return-value` throw an error if the call threw an error, since that's exactly what will happen if you call it on a spy-context-error struct. However, we should probably not implement it in this way, because we probably want to provide our own error message.

Of course, my proposal means we need to somehow redefine `spy-context-return-value` after the struct is defined. I guess we can do that with advice. Having a package define advice for its own function is kind of odd, but it sort of makes sense here, given that I don't think there's any other way to reliably redefine a struct accessor function.

Reply:
Regardless, I think I'm even further convinced that the only reasonable thing to do is to throw an error if a test tries to access the return value of a failed call. This will cause the test to fail, which is probably what we want if a function call is failing in a way that the test didn't expect.

Reply:
Any opinions on whether advising a function that was defined by `cl-defstruct` is an appropriate way to implement this? @snogge

Reply:
I'm not big on using advice this way, but I guess it is the least complicated way forward.

I still think the code below could be tweaked to do more or less the same and give room for future expansion.  But maybe it's too complicated?
```emacs-lisp
(cl-defstruct spy-context args current-buffer)
(cl-defstruct (spy-context--return (:include spy-context)) value)
(cl-defstruct (spy-context--thrown (:include spy-context)) signal)
;; Not sure when cl-defgeneric was introduced
(defgeneric spy-context-return-value (context)
  (error "Context is not a return"))
(defmethod spy-context-return-value ((context spy-context--return))
  (spy-context--return-value context))
(defgeneric spy-context-thrown-signal (context)
  (error "Context is not an error"))
(defmethod spy-context-thrown-signal ((context spy-context--thrown))
  (spy-context--thrown-signal context))
```

Reply:
Yeah, that looks good. I'll try to find out when `cl-defgeneric` was added to see if we can use it or not.

Reply:
Looks like `cl-defgeneric` was added in 25.1. I guess that means we can't use it if we're supporting all the way back to 24.3.

Reply:
Ok, this is now implemented using your suggestions, including `spy-calls-count-returned` and `spy-calls-count-errors`.